File: src/read.ts
import { parquetMetadataAsync, parquetSchema } from './metadata.js'
import { parquetPlan, prefetchAsyncBuffer } from './plan.js'
import { assembleAsync, asyncGroupToRows, readRowGroup } from './rowgroup.js'
import { concat } from './utils.js'
import {AsyncRowGroup, ParquetReadOptions, ParquetReadOptionsWithoutOnComplete, QueryPlan} from "./types.js";

/**
 * @import {AsyncRowGroup, DecodedArray, ParquetReadOptions, BaseParquetReadOptions} from '../src/types.js'
 */
/**
 * Read parquet data rows from a file-like object.
 * Reads the minimal number of row groups and columns to satisfy the request.
 *
 * Returns a void promise when complete.
 * Errors are thrown on the returned promise.
 * Data is returned in callbacks onComplete, onChunk, onPage, NOT the return promise.
 * See parquetReadObjects for a more convenient API.
 *
 * @param {ParquetReadOptions} options read options
 * @returns {Promise<void>} resolves when all requested rows and columns are parsed, all errors are thrown here
 */
export async function parquetRead(options: ParquetReadOptions): Promise<void> {
  // load metadata if not provided
  if (options.metadata === undefined) {
    options.metadata = await parquetMetadataAsync(options.file)
  }

  // read row groups
  const asyncGroups = parquetReadAsync(options)
  const rowStart = 0
  const rowEnd = Infinity
  const { onChunk, onComplete, rowFormat } = options

  // skip assembly if no onComplete or onChunk, but wait for reading to finish
  if (onComplete === undefined && onChunk === undefined) {
    for (const { asyncColumns } of asyncGroups) {
      for (const { data } of asyncColumns) await data
    }
    return
  }

  // assemble struct columns
  const schemaTree = parquetSchema(options.metadata.schema)
  const assembled = asyncGroups.map(arg => assembleAsync(arg, schemaTree))

  // onChunk emit all chunks (don't await)
  if (onChunk !== undefined) {
    for (const asyncGroup of assembled) {
      for (const asyncColumn of asyncGroup.asyncColumns) {
        asyncColumn.data.then(columnDatas => {
          let rowStart = asyncGroup.groupStart
          for (const columnData of columnDatas) {
            onChunk({
              columnName: asyncColumn.pathInSchema[0],
              columnData,
              rowStart,
              rowEnd: rowStart + columnData.length,
            })
            rowStart += columnData.length
          }
        })
      }
    }
  }

  // onComplete transpose column chunks to rows
  if (onComplete !== undefined) {
    // loosen the types to avoid duplicate code
    const rows: any[] = []
    for (const asyncGroup of assembled) {
      // filter to rows in range
      const selectStart = Math.max(rowStart - asyncGroup.groupStart, 0)
      const selectEnd = Math.min((rowEnd !== undefined ? rowEnd : Infinity) - asyncGroup.groupStart, asyncGroup.groupRows)
      // transpose column chunks to rows in output
      const groupData = rowFormat === 'object' ?
        await asyncGroupToRows(asyncGroup.asyncColumns, selectStart, selectEnd, 'object') :
        await asyncGroupToRows(asyncGroup.asyncColumns, selectStart, selectEnd, 'array')
      concat(rows, groupData)
    }
    onComplete(rows)
  } else {
    // wait for all async groups to finish (complete takes care of this)
    for (const { asyncColumns } of assembled) {
      for (const { data } of asyncColumns) await data
    }
  }
}

export function parquetReadAsync(options: ParquetReadOptions): AsyncRowGroup[] {
  if (!options.metadata) throw new Error('parquet requires metadata')
  // TODO: validate options (start, end, columns, etc)

  // prefetch byte ranges
  const plan: QueryPlan = parquetPlan(options.metadata)
  options.file = prefetchAsyncBuffer(options.file, plan.fetches)

  // read row groups
  return plan.groups.map(groupPlan => readRowGroup(options, plan.metadata, groupPlan))
}

/**
 * This is a helper function to read parquet row data as a promise.
 * It is a wrapper around the more configurable parquetRead function.
 *
 * @param {ParquetReadOptionsWithoutOnComplete} options
 * @returns {Promise<Record<string, any>[]>} resolves when all requested rows and columns are parsed
 */
export function parquetReadObjects(options: ParquetReadOptionsWithoutOnComplete): Promise<Record<string, any>[]> {
  return new Promise((onComplete, reject) => {
    parquetRead({
      ...options,
      rowFormat: 'object', // force object output
      onComplete,
    }).catch(reject)
  })
}

File: src/main/kotlin/app/softwork/hyparquet/Read.kt
package app.softwork.hyparquet

/**
 * Read parquet data rows from a file-like object.
 * Reads the minimal number of row groups and columns to satisfy the request.
 *
 * Returns a void when complete.
 * Errors are thrown.
 * Data is returned in callbacks onComplete, onChunk, onPage, NOT the return.
 *
 * @param options read options
 */
suspend fun parquetRead(options: ParquetReadOptions) {
    // load metadata if not provided
    val metadata = options.metadata ?: parquetMetadataAsync(options.file)

    // read row groups
    val asyncGroups = parquetReadAsync(options.copy(metadata = metadata))
    val rowStart = 0
    val rowEnd = Int.MAX_VALUE
    val onChunk = options.onChunk
    val onComplete = options.onComplete

    // skip assembly if no onComplete or onChunk, but wait for reading to finish
    if (onComplete == null && onChunk == null) {
        for (group in asyncGroups) {
            for (column in group.asyncColumns) {
                column.data()
            }
        }
        return
    }

    // assemble struct columns
    val schemaTree = parquetSchema(metadata.schema)
    val assembled = asyncGroups.map { assembleAsync(it, schemaTree) }

    // onChunk emit all chunks (don't await)
    if (onChunk != null) {
        for (asyncGroup in assembled) {
            for (asyncColumn in asyncGroup.asyncColumns) {
                val columnDatas = asyncColumn.data()
                var rowStart = asyncGroup.groupStart
                for (columnData in columnDatas) {
                    onChunk(ColumnData(
                        columnName = asyncColumn.pathInSchema[0],
                        columnData = columnData,
                        rowStart = rowStart,
                        rowEnd = rowStart + getArrayLength(columnData)
                    ))
                    rowStart += getArrayLength(columnData)
                }
            }
        }
    }

    // onComplete transpose column chunks to rows
    if (onComplete != null) {
        val rows = mutableListOf<Any?>()
        for (asyncGroup in assembled) {
            // filter to rows in range
            val selectStart = maxOf(rowStart - asyncGroup.groupStart, 0)
            val selectEnd = minOf(rowEnd - asyncGroup.groupStart, asyncGroup.groupRows)
            // transpose column chunks to rows in output
            val groupData = asyncGroupToRows(asyncGroup.asyncColumns, selectStart, selectEnd, options.rowFormat)
            concat(rows, DecodedArray.AnyArrayType(groupData))
        }
        @Suppress("UNCHECKED_CAST")
        onComplete(rows as List<Any>)
    } else {
        // wait for all async groups to finish (complete takes care of this)
        for (asyncGroup in assembled) {
            for (asyncColumn in asyncGroup.asyncColumns) {
                asyncColumn.data()
            }
        }
    }
}

fun parquetReadAsync(options: ParquetReadOptions): List<AsyncRowGroup> {
    val metadata = options.metadata ?: throw Error("parquet requires metadata")

    // prefetch byte ranges
    val plan = parquetPlan(metadata)
    val file = prefetchAsyncBuffer(options.file, plan.fetches)

    // read row groups
    return plan.groups.map { groupPlan -> readRowGroup(options.copy(file = file), metadata, groupPlan) }
}

/**
 * This is a helper function to read parquet row data as a promise.
 * It is a wrapper around the more configurable parquetRead function.
 *
 * @param options
 * @returns resolves when all requested rows and columns are parsed
 */
suspend fun parquetReadObjects(options: BaseParquetReadOptions): List<Map<String, Any?>> {
    val result = mutableListOf<Any>()
    parquetRead(ParquetReadOptions(
        file = options.file,
        metadata = options.metadata,
        onChunk = options.onChunk,
        onPage = options.onPage,
        compressors = options.compressors,
        utf8 = options.utf8,
        parsers = options.parsers,
        rowFormat = RowFormat.OBJECT,
        onComplete = { rows -> result.addAll(rows) }
    ))
    @Suppress("UNCHECKED_CAST")
    return result as List<Map<String, Any?>>
}

// Helper function
private fun getArrayLength(array: DecodedArray): Int {
    return when (array) {
        is DecodedArray.ByteArrayType -> array.array.size
        is DecodedArray.IntArrayType -> array.array.size
        is DecodedArray.LongArrayType -> array.array.size
        is DecodedArray.FloatArrayType -> array.array.size
        is DecodedArray.DoubleArrayType -> array.array.size
        is DecodedArray.AnyArrayType -> array.array.size
    }
}
